### PR TITLE
Update WPCS to v3.0, and fix all reported Coding Standards issues

### PR DESCRIPTION
**Pull request by aristath**
## What?
Updates WPCS to v3.0

## Why?
A few days ago v3.0.0 of the WPCS package was released by @jrfnl.
More details about the package and v3.0 can be found on the post on https://make.wordpress.org/core/2023/08/21/wordpresscs-3-0-0-is-now-available/

Updating the package will ensure that Gutenberg is compatible with PHP versions greater than 8.x, and enforces stricter rules for consistency.

## How?

* Updated the package
* Ran `phpcbf` to fix some spacing issues and all the "simple" stuff
* Ran `phpcs` and manually fixed the remaining issues

## Testing Instructions

If automated tests pass we're good.

### Testing Instructions for Keyboard

Not required.

### CODE REVIEW

File: composer.json
Comment (jrfnl) [inline composer.json]: As per the [upgrade guide](https://github.com/WordPress/WordPress-Coding-Standards/wiki/Upgrade-Guide-to-WordPressCS-3.0.0-for-ruleset-maintainers), the `dealerdirect/phpcodesniffer-composer-installer` package should be removed from `require-dev` and it is recommended to also remove the `squizlabs/php_codesniffer` package in favour of letting the actual standards packages (WordPressCS, VariableAnalysis, PHPCompatibility) manage the PHPCS version.

Both those packages will still remain installed, but having these restrictions here is a version restrictions conflict-hell waiting to happen (and is currently already blocking the Composer PHPCS plugin from updating to the 1.0 version).

File: lib/compat/wordpress-6.2/html-api/class-wp-html-tag-processor.php
Comment (jrfnl) [inline lib/compat/wordpress-6.2/html-api/class-wp-html-tag-processor.php]: Should this maybe be changed to a `do - while` loop instead ?

Also note: while having an assignment in a condition is not a great idea, it is allowed in WP Core (exception in the Core ruleset - [old version](https://github.com/WordPress/wordpress-develop/blob/5a99ab468a2bae1c4660ec45dde908f3b1032041/phpcs.xml.dist#L321-L324) - [new version](https://github.com/WordPress/wordpress-develop/pull/5047/files#diff-05ae9cddcaec1e845771a7db224961439f83ef5939ec67d3a48744cb34d7e58bR232-R235)), so it could be considered to port that exception over to GB as well.

Another alternative would be to update the ignore annotation to the new sniff name.

Having said all that, the best solution is, of course, not to have an assignment in a condition in the first place, but I thought it only fair to make you aware of the options here.

File: lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php
Comment (jrfnl) [inline lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php]: Same comment as per https://github.com/WordPress/gutenberg/pull/53866/files#diff-9b0b96f086429280a9f988226c77db353802dadbf18552cc9cf55db9da1a04aeR2254 / `while` condition in `lib/compat/wordpress-6.2/html-api/class-wp-html-tag-processor.php` file.

File: phpunit/experimental/interactivity-api/directive-processing-test.php
Comment (jrfnl) [inline phpunit/experimental/interactivity-api/directive-processing-test.php]:
```suggestion
 * @subpackage Interactivity API
 *
 * @phpcs:disable Generic.Files.OneObjectStructurePerFile.MultipleFound
 */

```

I.e.: the `disable` comment can be placed in the file docblock, which will leave the file looking cleaner.

I'd still recommend to actually fix this at some point by moving the extraneous class(es) to their own (fxiture ?) file(s).

File: test/emptytheme/functions.php
Comment (jrfnl) [inline test/emptytheme/functions.php:23]: I'm curious what triggered this change ? As far as I can see, this is not coming from WPCS 3.0, so I wonder if it belongs in this PR.

File: test/gutenberg-test-themes/behaviors-enabled/index.php
Comment (jrfnl) [inline test/gutenberg-test-themes/behaviors-enabled/index.php]: If these (this one and the next below it) are files which should _never_ be included in a scan, you could consider adding an exception to the ruleset instead of a comment.

File: test/gutenberg-test-themes/emptyhybrid/functions.php
Comment (jrfnl) [inline test/gutenberg-test-themes/emptyhybrid/functions.php:26]: Same comment as I left for the `test/emptytheme/functions.php` file.

File: phpcs.xml.dist
Comment (jrfnl) [inline phpcs.xml.dist:1]: I'd recommend having a look at [this commit](https://github.com/WordPress/wordpress-develop/pull/5047/commits/87ecda371c0d317fc45244709798bdb5f9c4f811) in the WPCS PR for inspiration on tidying up the ruleset.

Comment (jrfnl) [inline phpcs.xml.dist:1]: Line 13-14:
```xml
<rule ref="WordPress.WP.I18n"/>
<config name="text_domain" value="gutenberg,default"/>
```

There is only one sniff which takes a text domain, so this should not be configured like this, but like so:
```xml
<rule ref="WordPress.WP.I18n">
	<properties>
		<property name="text_domain" type="array">
			<element value="gutenberg"/>
			<element value="default"/>
		</property>
	</properties>
</rule>
```

Ref: https://github.com/WordPress/WordPress-Coding-Standards/wiki/Customizable-sniff-properties#wordpresswpi18n-setting-your-text-domain

**Review by jrfnl — COMMENTED** [review]
Hi @aristath Thanks for setting this up! Looking good. Just left a few remarks inline for you to consider.

File: composer.json
Comment (aristath) [reply composer.json]: Fixed 👍

File: lib/compat/wordpress-6.2/html-api/class-wp-html-tag-processor.php
Comment (aristath) [reply lib/compat/wordpress-6.2/html-api/class-wp-html-tag-processor.php]: Thank you Juliette!
I went with the `do - while` option. Like you said,
> the best solution is, of course, not to have an assignment in a condition in the first place

 so I went with that 👍

File: lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php
Comment (aristath) [reply lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php]: Same as above, went with the `do - while` option

File: test/gutenberg-test-themes/behaviors-enabled/index.php
Comment (aristath) [reply test/gutenberg-test-themes/behaviors-enabled/index.php]: Added an exception in the ruleset file 👍

File: test/emptytheme/functions.php
Comment (aristath) [reply test/emptytheme/functions.php:23]: You are correct, technically this does not belong in this PR since from what I can see, It's also an issue on `trunk`.
However, in this PR I fixed all CS issues that were reported. I'll change the title of the PR to reflect that 👍

**Review by SergeyBiryukov — APPROVED** [review]

**Comment by aristath** [comment]
Thank you for the review @SergeyBiryukov!

I rebased the PR, and as soon as automated tests complete we can merge this 👍

**Comment by dmsnell** [comment]
@aristath sorry I didn't catch this before it was merged.

for HTML API code this is pulled from Core, so changes need to happen first in Core and then be pulled into Gutenberg. second, the change from a `while` to a `do` probably introduced a vulnerability by using values that aren't checked. we should be careful to ensure our style preference changes don't impact the runtime behavior of code.

I'd like to revert the HTML API changes from here if it's okay with you and then work backwards from Core. can you create a patch/Trac ticket in Core to address the issues you have? in the meantime I'd like to revert so that we don't start maintaining two separate copies of this code.

**Comment by aristath** [comment]
> @aristath sorry I didn't catch this before it was merged.
> 
> for HTML API code this is pulled from Core, so changes need to happen first in Core and then be pulled into Gutenberg. second, the change from a `while` to a `do` probably introduced a vulnerability by using values that aren't checked. we should be careful to ensure our style preference changes don't impact the runtime behavior of code.
> 
> I'd like to revert the HTML API changes from here if it's okay with you and then work backwards from Core. can you create a patch/Trac ticket in Core to address the issues you have? in the meantime I'd like to revert so that we don't start maintaining two separate copies of this code.

Of course, we can revert the changes to the HTML API in Gutenberg.
I think that WPCS 3.0 changes were already merged in Core, so this should have already been updated (if there are any commits remaining I'm sure they'll be merged soon).

**Comment by dmsnell** [comment]
> I think that WPCS 3.0 changes were already merged in Core, so this should have already been updated (if there are any commits remaining I'm sure they'll be merged soon).

@aristath unfortunately not, and the two repos have different style preferences, so again, unfortunately, we now have divergent code in Core and Gutenberg and it's quite a mess. _some_ of the changed introduced here have been applied to `trunk` in Core, but they haven't been applied to the `6.3` branch, meaning that when we pull the already-accepted and already-merged code from Core into Gutenberg, the 6.3 compat layer _should_ have styles that WPCS doesn't like while eventually the 6.4 compat layer _shouldn't_.

currently I'm unable to revert without overriding the CI job because it's rejecting the change on style preferences, which is unfortunate, because it's enforcing that bug 🙃

I've mentioned you on #54481 where I'm trying to sort it all out, but admittedly, it's very confusing.

**Pull request merged**
